### ISSUE TITLE
pyodbc.ProgrammingError on Complex Queries

### ISSUE THREAD

**Issue by fifothekid** [mkleehammer/pyodbc#458]
- Python: 2.7.15 x86
- pyodbc: 4.0.24
- OS: Windows 10/Ubuntu 16.04
- DB: MS SQL Server 2017
- driver: FreeTDS 1.0.9

### Issue

I have some complex T-SQL queries that have multiple commands like this:
```
     declare @fim date
			IF (OBJECT_ID('[dbo].[fnUTCDateTB]') IS NOT NULL)
				SET @fim = (select dateadd(day,-4,convert(date,DataLocal)) from [dbo].[fnUTCDateTB](GETDATE()))
			ELSE
				SET @fim = dateadd(day,-4,convert(date,getdate()))
			 SELECT CST.service_id as service_id
			     ,  CST.ID as customer_id
		    	 ,  CFG.ID as customer_config_id
		    	 ,  CASE cct.frequency_type
                  		WHEN 'DAY'   THEN DATEADD(DAY  , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'MONTH' THEN DATEADD(MONTH, cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'YEAR'  THEN DATEADD(YEAR , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                    END as next_birthday
		      FROM  CUSTOMER cst (NOLOCK)
```

which used to work with pymssql without any problem. Later we had to switch to PyODBC because we needed to connected to an AlwaysOn secondary database, and this required the ApplicationIntent=ReadOnly parameter. I configured it by compiling and installing the latest FreeTDS plugin with ODBC. I tested it with tsql and isql tools, in addition to MS SQL Management Studio 17 and JetBrains DataGrip, and the queries worked as expected. The problem only happens with my Python script using PyODBC. I get this error with complex queries:

>     first_row = cur.fetchone()
> pyodbc.ProgrammingError: No results.  Previous SQL was not a query.

If I test with short queries they would work even with @DECLARE statements, while long queries work only replacing when removing the @DECLARE statements and substituting all the variables inside the query with their values.

I hope that you can give me a tip on how I can solve that.

**Comment by gordthompson** [comment]
Can you try putting `SET NOCOUNT ON;` at the beginning of your code block?

**Comment by fifothekid** [comment]
I did try that and I know what kind of problems does it solve however in these cases it didn't help
It's strange because the exact same query works perfectly with database administration software but with PyODBC it only works after I start removing parts of the query. I kept on experimenting for hours and I couldn't reach any reasonable conclusion on what exactly is causing these errors other than the query being "complex".

**Comment by gordthompson** [comment]
As a possible workaround, or at least for diagnostic purposes, can you catch the ProgrammingError thrown by `cur.fetchone()`, do a `cur.nextset()`, and then try `cur.fetchone()` again? You could even try repeating that process until `cur.nextset()` returns `None`.

**Comment by fifothekid** [comment]
I put a breakpoint in the exception handler, launched the PyCharm evaluation windows and ran cur.nextset(), which returned false, while cur.fetchone() only returns errors

**Comment by gordthompson** [comment]
Hmm, okay. Just to clarify: You reported using "FreeTDS 1.0.9" but later you said that you are using "the latest FreeTDS plugin with ODBC", so that would be more like FreeTDS 1.00.98, right?

**Comment by fifothekid** [comment]
there is a typo in src\textenc.h
OPENC_RAW should be changed to OPTENC_RAW

**Comment by fifothekid** [comment]
After fixing the typo I'm getting on Ubuntu:
```
osboxes@osboxes:~/Documents/pyodbc$ python setup.py build
running build
running build_ext
building 'pyodbc' extension
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-nbjU53/python2.7-2.7.15~rc1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DPYODBC_VERSION=4.0.25b11 -I/usr/include/python2.7 -c src/buffer.cpp -o build/temp.linux-x86_64-2.7/src/buffer.o -Wno-write-strings
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-nbjU53/python2.7-2.7.15~rc1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DPYODBC_VERSION=4.0.25b11 -I/usr/include/python2.7 -c src/cnxninfo.cpp -o build/temp.linux-x86_64-2.7/src/cnxninfo.o -Wno-write-strings
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
x86_64-linux-gnu-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fno-strict-aliasing -Wdate-time -D_FORTIFY_SOURCE=2 -g -fdebug-prefix-map=/build/python2.7-nbjU53/python2.7-2.7.15~rc1=. -fstack-protector-strong -Wformat -Werror=format-security -fPIC -DPYODBC_VERSION=4.0.25b11 -I/usr/include/python2.7 -c src/connection.cpp -o build/temp.linux-x86_64-2.7/src/connection.o -Wno-write-strings
cc1plus: warning: command line option ‘-Wstrict-prototypes’ is valid for C/ObjC but not for C++
In file included from /usr/include/python2.7/Python.h:97:0,
                 from src/pyodbc.h:45,
                 from src/connection.cpp:11:
src/connection.cpp: In function ‘bool Connect(PyObject*, HDBC, bool, long int, Object&)’:
/usr/include/python2.7/stringobject.h:91:55: error: invalid cast from type ‘Object’ to type ‘PyStringObject*’
 #define PyString_AS_STRING(op) (((PyStringObject *)(op))->ob_sval)
                                                       ^
src/connection.cpp:97:30: note: in expansion of macro ‘PyString_AS_STRING’
                 szEncoding = PyString_AS_STRING(encBytes);
                              ^~~~~~~~~~~~~~~~~~
error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

```
And on Windows:
```
D:\m4u\pyodbc>python setup.py build
running build
running build_ext
building 'pyodbc' extension
C:\Users\FiFo\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -DPYODBC_VERSION=4.0.25b11 -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\include -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\PC /Tpsrc\buffer.cpp /Fobuild\temp.win32-2.7\Release\src\buffer.obj /Wall /wd4514 /wd4820 /wd4668 /wd4711 /wd4100 /wd4127 /wd4191
buffer.cpp
C:\Users\FiFo\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -DPYODBC_VERSION=4.0.25b11 -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\include -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\PC /Tpsrc\cnxninfo.cpp /Fobuild\temp.win32-2.7\Release\src\cnxninfo.obj /Wall /wd4514 /wd4820 /wd4668 /wd4711 /wd4100 /wd4127 /wd4191
cnxninfo.cpp
C:\Users\FiFo\AppData\Local\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\Bin\cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -DPYODBC_VERSION=4.0.25b11 -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\include -IC:\Users\FiFo\AppData\Local\Programs\Python\Python27\PC /Tpsrc\connection.cpp /Fobuild\temp.win32-2.7\Release\src\connection.obj /Wall /wd4514 /wd4820 /wd4668 /wd4711 /wd4100 /wd4127 /wd4191
connection.cpp
src\connection.cpp(97) : error C2440: 'type cast' : cannot convert from 'Object' to 'PyStringObject *'
        No user-defined-conversion operator available that can perform this conversion, or the operator
cannot be called
src\connection.cpp(97) : error C2227: left of '->ob_sval' must point to class/struct/union/generic type
error: command 'C:\\Users\\FiFo\\AppData\\Local\\Programs\\Common\\Microsoft\\Visual C++ for Python\\9.0\\VC\\Bin\\cl.exe' failed with exit status 2
```

**Comment by gordthompson** [comment]
@mkleehammer - Confirmed. The latest master (d8e5537c) builds successfully under Python3 but not under Python2.

https://github.com/mkleehammer/pyodbc/blob/d8e5537c04dae715b72c364fbffe2b2eb4810e2d/src/textenc.h#L124

> c:\users\gord\git\pyodbc\src\textenc.h(124) : error C2065: 'OPENC_RAW' : undeclared identifier

**Comment by gordthompson** [comment]
@fifothekid - Troubleshooting suggestion:

Perhaps try hard-coding the anonymous code block into a Python script and try running that, e.g.,

```python
sql = """\
declare @fim date
            IF (OBJECT_ID('[dbo].[fnUTCDateTB]') IS NOT NULL)
                SET @fim = (select dateadd(day,-4,convert(date,DataLocal)) from [dbo].[fnUTCDateTB](GETDATE()))
            ELSE
                SET @fim = dateadd(day,-4,convert(date,getdate()))
             SELECT CST.service_id as service_id
                 ,  CST.ID as customer_id
                 ,  CFG.ID as customer_config_id
                 ,  CASE cct.frequency_type
                        WHEN 'DAY'   THEN DATEADD(DAY  , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'MONTH' THEN DATEADD(MONTH, cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'YEAR'  THEN DATEADD(YEAR , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                    END as next_birthday
              FROM  CUSTOMER cst (NOLOCK)
"""
first_row = cur.execute(sql).fetchone()
```

If that works then you can concentrate on how bash might be "adjusting" the SQL command text when it is passed to a Python script as an argument.

**Comment by gordthompson** [comment]
@mkleehammer - Build error also reported as #461

**Comment by v-chojas** [comment]
Can you try msodbcsql17 driver? It is available for both Windows 10 and Ubuntu 16.04. This could be FreeTDS bug.

**Comment by fifothekid** [comment]
The MS SQL Server driver didn't work for me I don't remember why exactly but it was simply giving random errors.
I'll do a trace and test with both isql and tsql tools today

**Comment by fifothekid** [comment]
I tested with tracing on both using the Python script and using the tsql tool that comes with FreeTDS
The Python with PyODBC gave an error
tsql did not

Here are the logs:
FreeTDS log with PyODBC
https://pastebin.com/G8KRMMvp

ODBC log with PyODBC
https://pastebin.com/SWcAJKDR

FreeTDS log with tsql:
https://pastebin.com/hB4Xcbsa

Note that the tsql test does not include an ODBC log because it does not use ODBC. It passes the query directly to the FreeTDS driver
I tried doing the same tests with the isql tool that comes with unixODBC on Ubuntu but I could not find any way to pass multi-line queries to it. even with the \n parameter it would start processing partial queries and throwing errors.

**Comment by gordthompson** [comment]
@fifothekid - For what it's worth, the following works fine for my under Xubuntu 18.04:

```python
import pyodbc
import sys

print(sys.version)
# 2.7.15rc1 (default, Apr 15 2018, 21:51:34)
print(pyodbc.version)
# 4.0.24
conn_str = (
    r'DRIVER=FreeTDS;'
    r'SERVER=192.168.0.179;'
    r'PORT=49242;'
    r'DATABASE=issue458;'
    r'UID=sa;PWD=whatever;'
)
cnxn = pyodbc.connect(conn_str, autocommit=True)
print(cnxn.getinfo(pyodbc.SQL_DRIVER_NAME) + '  ' + cnxn.getinfo(pyodbc.SQL_DRIVER_VER))
# libtdsodbc.so  01.00.0098
cur = cnxn.cursor()
sql = """\
SET NOCOUNT ON;
declare @fim date
IF (OBJECT_ID('[dbo].[fnUTCDateTB]') IS NOT NULL)
    SET @fim = (select dateadd(day,-4,convert(date,DataLocal)) from [dbo].[fnUTCDateTB](GETDATE()))
ELSE
    SET @fim = dateadd(day,-4,convert(date,getdate()))
SELECT CST.service_id as service_id
    ,  CST.ID as customer_id
    ,  CFG.ID as customer_config_id
    ,  CASE cct.frequency_type
            WHEN 'DAY'   THEN DATEADD(DAY  , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
            WHEN 'MONTH' THEN DATEADD(MONTH, cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
            WHEN 'YEAR'  THEN DATEADD(YEAR , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
        END as next_birthday
FROM  CUSTOMER cst (NOLOCK)
    JOIN  CUSTOMER_CONFIG cfg (NOLOCK)
        ON  cfg.id = cst.customer_config_id
    JOIN  PROFILE prf (NOLOCK)
        ON  prf.id = cfg.profile_id
    JOIN  PROFILE_CONFIG pfg (NOLOCK)
        ON  pfg.id = prf.profile_config_id
    JOIN  CYCLE_TYPE cct (NOLOCK)
        ON  pfg.cycle_type_id = cct.id
where  cfg.active = 1
    and cfg.created <= @fim
    and CONVERT( DATE
            , dateadd( day
                , pfg.reattempt_limit_in_days
                , CASE cct.frequency_type
                        WHEN 'DAY'   THEN DATEADD(DAY  , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'MONTH' THEN DATEADD(MONTH, cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                        WHEN 'YEAR'  THEN DATEADD(YEAR , cct.frequency_quantity * cfg.cycle_number, cfg.birthday)
                  END)
        ) < @fim
"""
cur.execute(sql)
first_row = cur.fetchone()
print(first_row)
# (123, 1, 1, datetime.date(1957, 7, 8))

cur.close()
cnxn.close()
```

**Comment by fifothekid** [comment]
interesting...
I'll check this today thanks a lot

**Comment by fifothekid** [comment]
The CMake build script has only the basic options. there is no way to select MSDBLib nor to choose target TDS version
Later I'll build it on Ubuntu and try again

**Comment by v-chojas** [comment]
Looking at the TDS packet returned from the server when you attempt to execute that large query, it has several results which it seems FreeTDS isn't handling correctly.

**Comment by fifothekid** [comment]
but passing the same query to the tsql tool that comes with FreeTDS everything works as expected
can you please compare the two logs?

**Comment by gordthompson** [comment]
@fifothekid - A cursory inspection of the two TDS logs shows that `tsql` uses the FreeTDS *API* directly, while pyodbc uses the FreeTDS *ODBC driver* (which in turn calls the FreeTDS API). Example:

tsql:

```
token.c:1616: name                 size/wsize      type/wtype      utype  
token.c:1617: -------------------- --------------- --------------- -------
token.c:1622: service_id                 2/2            52/52            0
token.c:1622: customer_id                8/8           127/127           0
token.c:1622: customer_config_id         8/8           127/127           0
token.c:1622: next_birthday              8/8           111/111           0
util.c:165:Changed query state from READING to PENDING
token.c:547:tds_process_tokens(0022EC78, 00CFF744, 00000000, 0x6914)
util.c:165:Changed query state from PENDING to READING
token.c:562:processing result tokens.  marker is  fd(DONE)
token.c:2102:tds_process_end: more_results = 0
		was_cancelled = 0
		error = 0
		done_count_valid = 0
```

pyodbc:

```
token.c:1616: name                 size/wsize      type/wtype      utype  
token.c:1617: -------------------- --------------- --------------- -------
token.c:1622: service_id                 2/2            52/52            0
token.c:1622: customer_id                8/8           127/127           0
token.c:1622: customer_config_id         8/8           127/127           0
token.c:1622: next_birthday              8/8           111/111           0
util.c:165:Changed query state from READING to PENDING
odbc.c:3611:odbc_process_tokens: tds_process_tokens returned 0
odbc.c:3612:    result_type=4049, TDS_DONE_COUNT=0, TDS_DONE_ERROR=0
odbc.c:3684:odbc_process_tokens: returning result_type 4049
odbc.c:3433:_SQLExecute: odbc_process_tokens returned result_type 4049
token.c:547:tds_process_tokens(00962B68, 007EF508, 007EF4F8, 0x6914)
util.c:165:Changed query state from PENDING to READING
token.c:562:processing result tokens.  marker is  fd(DONE)
token.c:2102:tds_process_end: more_results = 0
		was_cancelled = 0
		error = 0
		done_count_valid = 0
```

Notice that the pyodbc trace includes entries for "odbc.c", but the tsql trace does not.

So it's quite possible that the ODBC driver part of FreeTDS may end up treating server responses slightly differently than the native FreeTDS API itself.

**Comment by fifothekid** [comment]
OK I've just tested simple and complex queries on isql tool in Ubuntu
Simple queries return 0 when there are no results, while complex ones return -1 (which is TDS_NO_COUNT)
How are you dealing with this case? Are you considering it as non-query?

**Comment by mkleehammer** [comment]
Does everyone think this is a FreeTDS issue, then?  Should we document the FreeTDS issue number and close?

**Comment by gordthompson** [comment]
Aha. 

> Simple queries return 0 **when there are no results**

My repro code was failing to fail because my test data *was* returning results. Same repro code [as above](https://github.com/mkleehammer/pyodbc/issues/458#issuecomment-423763047) with tweaked data reproduced the issue under FreeTDS 1.00.99.

[freetds_fail.log](https://github.com/mkleehammer/pyodbc/files/2430682/freetds_fail.log), starting at line 249

```
[ODBC][2030][1538227836.826322][SQLExecDirect.c][240]
        Entry:
            Statement = 0x15f8910
            SQL = [SET NOCOUNT ON;
declare @fim date
IF (OBJECT_ID('[dbo].[fnUTCDateTB]') IS NOT NULL)
    SET @fim = (select dateadd(day,-4,conver...][length = 1612]
[ODBC][2030][1538227836.841765][SQLExecDirect.c][521]
        Exit:[SQL_NO_DATA]
```

However, the same code runs without error using msodbcsql 17.

[msodbcsql_ok.log](https://github.com/mkleehammer/pyodbc/files/2430683/msodbcsql_ok.log), starting at line 265:

```
[ODBC][2047][1538228019.656875][SQLExecDirect.c][240]
        Entry:
            Statement = 0x143e970
            SQL = [SET NOCOUNT ON;
declare @fim date
IF (OBJECT_ID('[dbo].[fnUTCDateTB]') IS NOT NULL)
    SET @fim = (select dateadd(day,-4,conver...][length = 1612]
[ODBC][2047][1538228019.683265][SQLExecDirect.c][521]
        Exit:[SQL_SUCCESS]
[ODBC][2047][1538228019.683351][SQLRowCount.c][173]
        Entry:
            Statement = 0x143e970
            Row Count = 0x7fffbb9b3b80
[ODBC][2047][1538228019.683363][SQLRowCount.c][247]
        Exit:[SQL_SUCCESS]
            Row Count = 0x7fffbb9b3b80 -> 0
```

**Comment by fifothekid** [comment]
I couldn't get msodbcsql 17 to work in our environment. besides, I cannot really migrate away from FreeTDS send my team does not have the luxury of time and resource to implement and test that anymore. The FreeTDS implementation is working if we just keep the queries simple

**Comment by gordthompson** [comment]
I only included the results from ODBC Driver 17 for SQL Server to illustrate that the issue is with the inconsistent behaviour of the FreeTDS ODBC driver and not a problem with pyodbc *per se*.

Of course, you can work around the issue with something like

```python
first_row = None
try:
    first_row = cur.fetchone()
except pyodbc.ProgrammingError as pe:
    if pe.message != 'No results.  Previous SQL was not a query.':
        raise
```

but the ultimate fix would have to come from FreeTDS.

So in response to the comment from @mkleehammer - yes, I think we can close this issue now.

**Comment by fifothekid** [comment]
I'll continue this conversation with the FreeTDS team
Thanks a lot!

**Comment by fifothekid** [comment]
I can confirm that the bug was fixed with FreeTDS version 1.00.104.
